libyul/backends/evm/StackHelpers.h

```diff
@@ -377,6 +377,147 @@ class Shuffler
+		m_junkSlotMultiplicity(0) {}
+
+	int& operator[](StackSlot const& _slot)
+	{
+		if (auto* p = std::get_if<FunctionCallReturnLabelSlot>(&_slot))
+			return m_functionCallReturnLabelSlotMultiplicity[*p];
+		if (std::holds_alternative<FunctionReturnLabelSlot>(_slot))
+			return m_functionReturnLabelSlotMultiplicity;
+		if (auto* p = std::get_if<VariableSlot>(&_slot))
+			return m_variableSlotMultiplicity[*p];
+		if (auto* p = std::get_if<LiteralSlot>(&_slot))
+			return m_literalSlotMultiplicity[*p];
+		if (auto* p = std::get_if<TemporarySlot>(&_slot))
+			return m_temporarySlotMultiplicity[*p];
+		return m_junkSlotMultiplicity;
+	}
+
+	int at(StackSlot const& _slot) const
+	{
+		if (auto* p = std::get_if<FunctionCallReturnLabelSlot>(&_slot))
+			return m_functionCallReturnLabelSlotMultiplicity.at(*p);
+		if (std::holds_alternative<FunctionReturnLabelSlot>(_slot))
+			return m_functionReturnLabelSlotMultiplicity;
+		if (auto* p = std::get_if<VariableSlot>(&_slot))
+			return m_variableSlotMultiplicity.at(*p);
+		if (auto* p = std::get_if<LiteralSlot>(&_slot))
+			return m_literalSlotMultiplicity.at(*p);
+		if (auto* p = std::get_if<TemporarySlot>(&_slot))
+			return m_temporarySlotMultiplicity.at(*p);
+		return m_junkSlotMultiplicity;
+	}
+
+private:
+	template <class T>
+	class StackSlotHashMap
+	{
+	public:
+		StackSlotHashMap(size_t _initialCapacity)
+		{
+			m_capacity = 0;
+			m_size = 0;
+			m_initialCapacity = 2;
+			while (m_initialCapacity < _initialCapacity)
+				m_initialCapacity *= 2;
+		}
+
+		int& operator[](T const& _key)
+		{
+			if (m_capacity == 0)
+			{
+				m_capacity = m_initialCapacity;
+				m_elements.resize(m_initialCapacity);
+			}
+
+			++m_size;
+			// Use a 50% load factor for simplicity and performance.
+			if (m_size > m_capacity / 2)
+			{
+				m_capacity *= 2;
+				std::vector<Element> newElements(m_capacity);
+				for (auto const& element: m_elements)
+					if (element.first)
+					{
+						size_t i = hash(element.first);
+						while (newElements[i].first)
+							i = inc(i);
+						newElements[i] = element;
+					}
+				m_elements = std::move(newElements);
+			}
+			uintptr_t id = _key.id();
+			for (size_t i = hash(id); ; i = inc(i))
+			{
+				if (!m_elements[i].first)
+				{
+					m_elements[i].first = id;
+					return m_elements[i].second;
+				}
+				if (m_elements[i].first == id)
+					return m_elements[i].second;
+			}
+		}
+
+		int at(T const& _key) const
+		{
+			if (m_capacity == 0)
+				return notFound();
+			uintptr_t id = _key.id();
+			for (size_t i = hash(id); ; i = inc(i))
+			{
+				if (!m_elements[i].first)
+					return notFound();
+				if (m_elements[i].first == id)
+					return m_elements[i].second;
+			}
+		}
+
+	private:
+		int notFound() const
+		{
+			throw std::out_of_range("Multiplicity::at() out of range");
+			return 0;
+		}
+
+		size_t hash(uintptr_t _id) const
+		{
+			_id ^= (_id >> 3) * 16777619;
+			_id ^= (_id >> 11) * 2166136261;
+			return _id & (m_capacity - 1);
+		}
```

Comment:
Who wrote this hash function? GPT-4? :)

Reply:
Yes. It is made for hashing pointers, which are often multiples of 8,16,32.

Reply:
I’ve written some Python snippet to verify that it indeed does distribute well for the intended use case.